scenario/state.py

```diff
@@ -165,67 +195,184 @@ class Relation(_DCBase):
         """Sugar to generate a <this relation>-relation-created event."""
         return Event(
             name=normalize_name(self.endpoint + "-relation-created"), relation=self
         )
 
     @property
-    def departed_event(self):
+    def departed_event(self) -> "Event":
         """Sugar to generate a <this relation>-relation-departed event."""
         return Event(
             name=normalize_name(self.endpoint + "-relation-departed"), relation=self
         )
 
     @property
-    def broken_event(self):
+    def broken_event(self) -> "Event":
         """Sugar to generate a <this relation>-relation-broken event."""
         return Event(
             name=normalize_name(self.endpoint + "-relation-broken"), relation=self
         )
 
 
+def unify_ids_and_remote_units_data(ids: List[int], data: Dict[int, Any]):
+    """Unify and validate a list of unit IDs and a mapping from said ids to databag contents.
+
+    This allows the user to pass equivalently:
+    ids = []
+    data = {1: {}}
+
+    or
+
+    ids = [1]
+    data = {}
+
+    or
+
+    ids = [1]
+    data = {1: {}}
+
+    but catch the inconsistent:
+
+    ids = [1]
+    data = {2: {}}
+
+    or
+
+    ids = [2]
+    data = {1: {}}
+    """
+    if ids and data:
+        if not set(ids) == set(data):
+            raise StateValidationError(
+                f"{ids} should include any and all IDs from {data}"
+            )
+    elif ids:
+        data = {x: {} for x in ids}
+    elif data:
+        ids = [x for x in data]
+    else:
+        ids = [0]
+        data = {0: {}}
+    return ids, data
+
+
+@dataclasses.dataclass
+class Relation(RelationBase):
+    __type__ = RelationType.regular
+    remote_app_name: str = "remote"
+    remote_unit_ids: List[int] = dataclasses.field(default_factory=list)
+
+    # local limit
+    limit: int = 1
+
+    remote_app_data: Dict[str, str] = dataclasses.field(default_factory=dict)
+    remote_units_data: Dict[int, Dict[str, str]] = dataclasses.field(
+        default_factory=dict
+    )
+
+    @property
+    def __databags__(self):
+        yield self.local_app_data
+        yield self.local_unit_data
+        yield self.remote_app_data
+        yield from self.remote_units_data.values()
+
+    def __post_init__(self):
+        super().__post_init__()
+        self.remote_unit_ids, self.remote_units_data = unify_ids_and_remote_units_data(
+            self.remote_unit_ids, self.remote_units_data
+        )
+
+
+@dataclasses.dataclass
+class SubordinateRelation(RelationBase):
+    __type__ = RelationType.subordinate
+
+    # todo: consider renaming them to primary_*_data
+    remote_app_data: Dict[str, str] = dataclasses.field(default_factory=dict)
+    remote_unit_data: Dict[str, str] = dataclasses.field(default_factory=dict)
+
+    # app name and ID of the primary that *this unit* is attached to.
+    primary_app_name: str = "remote"
+    primary_id: int = 0
+
+    @property
+    def __databags__(self):
```

Comment:
Why dunder? Perhaps `_databags` or [`__databags`](https://peps.python.org/pep-0008/#method-names-and-instance-variables)?